File: src/clients/javascript/test_phantomdb.js
/**
 * Unit tests for the PhantomDB JavaScript client.
 */

const { PhantomDB, PhantomDBError } = require('./phantomdb.js');

// Mock fetch for testing
global.fetch = jest.fn();

describe('PhantomDB', () => {
  let db;
  
  beforeEach(() => {
    db = new PhantomDB('http://localhost:8080');
    fetch.mockClear();
  });
  
  describe('healthCheck', () => {
    it('should check server health', async () => {
      // Mock response
      const mockResponse = {
        ok: true,
        text: async () => JSON.stringify({ status: 'healthy' })
      };
      
      fetch.mockResolvedValueOnce(mockResponse);
      
      // Call the method
      const result = await db.healthCheck();
      
      // Assertions
      expect(result).toEqual({ status: 'healthy' });
      expect(fetch).toHaveBeenCalledWith(
        'http://localhost:8080/health',
        {
          method: 'GET',
          headers: {
            'Content-Type': 'application/json'
          }
        }
      );
    });
  });
  
  describe('listDatabases', () => {
    it('should list databases', async () => {
      // Mock response
      const mockResponse = {
        ok: true,
        text: async () => JSON.stringify({ databases: ['db1', 'db2'] })
      };
      
      fetch.mockResolvedValueOnce(mockResponse);
      
      // Call the method
      const result = await db.listDatabases();
      
      // Assertions
      expect(result).toEqual(['db1', 'db2']);
      expect(fetch).toHaveBeenCalledWith(
        'http://localhost:8080/databases',
        {
          method: 'GET',
          headers: {
            'Content-Type': 'application/json'
          }
        }
      );
    });
  });
  
  describe('createDatabase', () => {
    it('should create a database', async () => {
      // Mock response
      const mockResponse = {
        ok: true,
        text: async () => JSON.stringify({ message: 'Database created' })
      };
      
      fetch.mockResolvedValueOnce(mockResponse);
      
      // Call the method
      const result = await db.createDatabase('testdb');
      
      // Assertions
      expect(result).toEqual({ message: 'Database created' });
      expect(fetch).toHaveBeenCalledWith(
        'http://localhost:8080/databases',
        {
          method: 'POST',
          headers: {
            'Content-Type': 'application/json'
          },
          body: JSON.stringify({ name: 'testdb' })
        }
      );
    });
  });
  
  describe('error handling', () => {
    it('should handle request exceptions', async () => {
      // Mock exception
      fetch.mockRejectedValueOnce(new Error('Connection failed'));
      
      // Assertions
      await expect(db.healthCheck()).rejects.toThrow(PhantomDBError);
    });
    
    it('should handle HTTP errors', async () => {
      // Mock HTTP error
      const mockResponse = {
        ok: false,
        status: 500,
        statusText: 'Internal Server Error'
      };
      
      fetch.mockResolvedValueOnce(mockResponse);
      
      // Assertions
      await expect(db.healthCheck()).rejects.toThrow(PhantomDBError);
    });
  });
});

// If running directly (not in Jest), run a simple test
if (typeof jest === 'undefined') {
  console.log('Running simple test...');
  
  // Mock fetch for simple test
  global.fetch = async (url, options) => {
    if (url === 'http://localhost:8080/health') {
      return {
        ok: true,
        text: async () => JSON.stringify({ status: 'healthy' })
      };
    }
    throw new Error('Unexpected URL');
  };
  
  async function runSimpleTest() {
    const db = new PhantomDB('http://localhost:8080');
    try {
      const health = await db.healthCheck();
      console.log('Simple test passed:', health);
    } catch (error) {
      console.error('Simple test failed:', error.message);
    }
  }
  
  runSimpleTest();
}

File: src/clients/test_phantomdb.py
"""
Unit tests for the PhantomDB Python client.
"""

import unittest
from unittest.mock import patch, Mock
import json
from phantomdb import PhantomDB, PhantomDBError


class TestPhantomDB(unittest.TestCase):
    """Test cases for the PhantomDB client."""
    
    def setUp(self):
        """Set up test fixtures before each test method."""
        self.db = PhantomDB("http://localhost:8080")
    
    @patch('phantomdb.requests.Session.request')
    def test_health_check(self, mock_request):
        """Test health check functionality."""
        # Mock response
        mock_response = Mock()
        mock_response.json.return_value = {"status": "healthy"}
        mock_response.content = json.dumps({"status": "healthy"}).encode()
        mock_response.raise_for_status.return_value = None
        mock_request.return_value = mock_response
        
        # Call the method
        result = self.db.health_check()
        
        # Assertions
        self.assertEqual(result, {"status": "healthy"})
        mock_request.assert_called_once_with("GET", "http://localhost:8080/health")
    
    @patch('phantomdb.requests.Session.request')
    def test_list_databases(self, mock_request):
        """Test listing databases."""
        # Mock response
        mock_response = Mock()
        mock_response.json.return_value = {"databases": ["db1", "db2"]}
        mock_response.content = json.dumps({"databases": ["db1", "db2"]}).encode()
        mock_response.raise_for_status.return_value = None
        mock_request.return_value = mock_response
        
        # Call the method
        result = self.db.list_databases()
        
        # Assertions
        self.assertEqual(result, ["db1", "db2"])
        mock_request.assert_called_once_with("GET", "http://localhost:8080/databases")
    
    @patch('phantomdb.requests.Session.request')
    def test_create_database(self, mock_request):
        """Test creating a database."""
        # Mock response
        mock_response = Mock()
        mock_response.json.return_value = {"message": "Database created"}
        mock_response.content = json.dumps({"message": "Database created"}).encode()
        mock_response.raise_for_status.return_value = None
        mock_request.return_value = mock_response
        
        # Call the method
        result = self.db.create_database("testdb")
        
        # Assertions
        self.assertEqual(result, {"message": "Database created"})
        mock_request.assert_called_once_with(
            "POST", 
            "http://localhost:8080/databases",
            json={"name": "testdb"},
            headers={'Content-Type': 'application/json'}
        )
    
    @patch('phantomdb.requests.Session.request')
    def test_request_exception(self, mock_request):
        """Test handling of request exceptions."""
        # Mock exception
        mock_request.side_effect = Exception("Connection failed")
        
        # Assertions
        with self.assertRaises(PhantomDBError):
            self.db.health_check()


if __name__ == '__main__':
    unittest.main()